proc_macros/src/lib.rs

```diff
@@ -1,801 +1,34 @@
-            if !content.is_empty() {
-                content.parse::<Token![,]>()?;
-            }
-        }
-        Ok(Self {
-            code: code.expect("Expected 'code'"),
-            errors: errors.expect("Expected 'errors'"),
-            output,
-        })
-    }
-}
-
-impl ToTokens for InvalidRuleTestSpec {
-    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
-        let code = &self.code;
-        let errors = &self.errors;
-        let output = match self.output.as_ref() {
-            Some(output) => quote! {
-                Some(#output)
-            },
-            None => quote! {
-                Option::<String>::None
-            },
-        };
-        quote! {
-            crate::RuleTestInvalid::new(
-                #code,
-                #errors,
-                #output
-            )
-        }
-        .to_tokens(tokens)
-    }
-}
-
-struct RuleTests {
-    valid: ExprArray,
-    invalid: Vec<InvalidRuleTestSpec>,
-}
-
-impl Parse for RuleTests {
-    fn parse(input: ParseStream) -> syn::Result<Self> {
-        let mut valid: Option<ExprArray> = Default::default();
-        let mut invalid: Option<Vec<InvalidRuleTestSpec>> = Default::default();
-        while !input.is_empty() {
-            let key: Ident = input.parse()?;
-            input.parse::<Token![=>]>()?;
-            match &*key.to_string() {
-                "valid" => {
-                    assert!(valid.is_none(), "Already saw 'valid' key");
-                    valid = Some(input.parse()?);
-                }
-                "invalid" => {
-                    assert!(invalid.is_none(), "Already saw 'invalid' key");
-                    let invalid_content;
-                    bracketed!(invalid_content in input);
-                    let invalid = invalid.get_or_insert_with(|| Default::default());
-                    while !invalid_content.is_empty() {
-                        let invalid_rule_test_spec: InvalidRuleTestSpec =
-                            invalid_content.parse()?;
-                        invalid.push(invalid_rule_test_spec);
-                        if !invalid_content.is_empty() {
-                            invalid_content.parse::<Token![,]>()?;
-                        }
-                    }
-                }
-                _ => panic!("didn't expect key '{}'", key),
-            }
-            if !input.is_empty() {
-                input.parse::<Token![,]>()?;
-            }
-        }
-        Ok(Self {
-            valid: valid.expect("Expected 'valid'"),
-            invalid: invalid.expect("Expected 'invalid'"),
-        })
-    }
+    builder_args::builder_args(input)
 }
 
 #[proc_macro]
 pub fn rule_tests(input: TokenStream) -> TokenStream {
-    let RuleTests { valid, invalid } = parse_macro_input!(input);
-
-    quote! {
-        crate::RuleTests::new(
-            #valid,
-            vec![#(#invalid),*],
-        )
-    }
-    .into()
+    rule_tests::rule_tests(input, "tree_sitter_lint")
```

Comment:
I think the only actual change is that per a previous PR comment when you use these macros from inside `tree_sitter_lint` it doesn't seem happy about referring to itself as `tree_sitter_lint` (vs `crate`)

So now we want to be able to invoke these macros from both "internal" and "external" code so there may be a better way but I just basically parameterized via "what should the "self crate" name be"

Reply:
I can't remember what it is, but I think there is some special token you can use to do this. `$crate` or something like that?

Reply:
Ok ya I think that crossed my mind at some point and I decided (without actually trying it) that "as opposed to declarative macros which can be defined in the "main" crate, for proc macros that wouldn't work because they have to go in their own crate and therefore `$crate` must resolve to the proc-macro crate name not the real "parent" crate name"?